src/librustc_middle/ty/query/on_disk_cache.rs

```diff
@@ -371,13 +371,13 @@ impl<'sess> OnDiskCache<'sess> {
         let pos = index.get(&dep_node_index).cloned()?;
 
         // Initialize `cnum_map` using the value from the thread that finishes the closure first.
-        self.cnum_map.init_nonlocking_same(|| Self::compute_cnum_map(tcx, &self.prev_cnums[..]));
```

Comment:
Only use of optimistic initialization strategy.

Reply:
I believe this should be replaceable with something like this:

```rust
if self.cnum_map.get().is_none() {
    self.cnum_map.set(Self::compute_cnum_map(...)));
}
```

Reply:
Indeed! Do you see a reason why we don't just lock here though? I believe @Zoxc authored this originally, but they are on hiatus.

Reply:
AFAICT this is called for every single query result we deserialize from disk -- that means likely hundreds of thousands of calls, if not more, and in quite hot code -- I imagine locking may be quite expensive as such. I'm not sure why we're unable to initially the crate number map earlier, though; I would expect there to be a point in between query deserialization starting and having the prev_cnums available.

Reply:
Sounds good. I'll switch to your suggestion and leave a note.

Reply:
To be clear, `OnceCell` already has a fast path that doesn't take the mutex when the value is already set. The only difference in behavior would be that only a single thread executes `compute_cnum_map` while the rest get de-scheduled. This seems like it might actually be an improvement over the status quo.

Reply:
Oh in that case yeah this is fine to leave as is, I wasn't sure if that optimization was built-in yet. I can't imagine it matters whether we wait on a mutex or not ~N times when N is so low (on powerful hardware today, probably only a hundred or so at most).